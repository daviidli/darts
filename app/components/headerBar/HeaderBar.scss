.header {
	padding: 0 2em;
}
