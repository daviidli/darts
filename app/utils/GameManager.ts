/* eslint-disable no-await-in-loop */
import DartBoard from './dartBoard';
import store from '../store/store';
import {
	setCurrentRound,
	setCurrentPlayer,
	initializedRound,
	setWinner,
	setWaiting,
	setMaxDarts,
	setMaxRounds,
	setCurrentDart,
	setTotals
} from '../actions/gameActions';

abstract class GameManager {
	private dartboard: DartBoard;
	private running: boolean;
	private missedDart: boolean;
	private dartsCleared: boolean;

	constructor() {
		const state = store.getState();

		if (!state.serialPort.length) {
			throw new Error('No serial port selected.');
		}

		this.dartboard = new DartBoard(state.serialPort);
		this.running = false;
		this.missedDart = false;
		this.dartsCleared = false;
	}

	protected abstract initialize(): void;
	protected abstract async turn(): Promise<void>;

	protected async run() {
		const {
			maxRounds,
			maxDarts,
			players,
			clearDartsWaitTime
		} = store.getState();

		const roundsArr = this.createStrArray(
			players.length,
			maxRounds,
			maxDarts
		);
		store.dispatch(initializedRound(roundsArr));

		this.running = true;
		for (let r = 0; r < maxRounds && this.running; r++) {
			store.dispatch(setCurrentRound(r));
			for (let p = 0; p < players.length && this.running; p++) {
				store.dispatch(setCurrentPlayer(p));
				await this.turn();
				await this.waitForClearingDarts(clearDartsWaitTime);
			}
		}
	}

	protected async getDart(): Promise<string> {
		let currentDart;
		while (true) {
			try {
				currentDart = await this.dartboard.getResults(1, 500);
			} catch (e) {
				// continue regardless of error
				// todo: show error notification
			}

			if (this.missedDart) {
				this.missedDart = false;
				return 'miss';
			}

			if (currentDart) {
				return currentDart[0];
			}
		}
	}

	public start() {
		this.initialize();
		this.run();
	}

	public stop() {
		this.running = false;
		this.dartboard.close();
		store.dispatch(setMaxDarts(0));
		store.dispatch(setMaxRounds(0));
		store.dispatch(setCurrentPlayer(0));
		store.dispatch(setCurrentDart(0));
		store.dispatch(setCurrentRound(0));
		store.dispatch(initializedRound([]));
		store.dispatch(setTotals([]));
	}

	public setWinner(index: number) {
		this.stop();
		store.dispatch(setWinner(index));
	}

	public miss() {
		this.missedDart = true;
	}

	public getRunning() {
		return this.running;
	}

	public clearedDarts() {
		this.dartsCleared = true;
	}

	private async waitForClearingDarts(timeout = 0) {
		this.dartsCleared = false;
		store.dispatch(setWaiting(true));
		let timeouts = 0;

		while (true) {
			await this.timeout(1000);
			timeouts += 1;
			if (timeouts === timeout) {
				store.dispatch(setWaiting(false));
				return;
			}
			if (this.dartsCleared) {
				this.dartsCleared = false;
				store.dispatch(setWaiting(false));
				return;
			}
		}
	}

	private timeout(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	private createStrArray(
		iMax: number,
		jMax: number,
		kMax: number
	): string[][][] {
		const initialRounds: string[][][] = [];
		for (let i = 0; i < iMax; i++) {
			const arr = [];
			for (let j = 0; j < jMax; j++) {
				arr.push(new Array(kMax).fill(''));
			}
			initialRounds.push(arr);
		}
		return initialRounds;
	}
}

export default GameManager;
