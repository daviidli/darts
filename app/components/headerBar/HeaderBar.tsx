import React from 'react';
import { Grid, Button } from '@material-ui/core';
import { useHistory } from 'react-router-dom';
import GameManager from '../../utils/GameManager';
import styles from './HeaderBar.scss';
import routes from '../../constants/routes.json';

export type Props = {
	gameManager: GameManager;
};

const HeaderBar = (props: Props) => {
	const { gameManager } = props;
	const history = useHistory();

	const goBack = () => {
		gameManager.stop();
		history.push(routes.HOME);
	};

	return (
		<Grid
			container
			direction="row"
			justify="space-around"
			alignItems="center"
			className={styles.header}
		>
			<Grid item xs>
				<Button variant="outlined" onClick={goBack}>
					End Game
				</Button>
			</Grid>
			<Grid item xs container justify="flex-end">
				<Button
					variant="contained"
					color="secondary"
					onClick={() => gameManager.miss()}
				>
					Miss
				</Button>
			</Grid>
		</Grid>
	);
};

export default HeaderBar;
