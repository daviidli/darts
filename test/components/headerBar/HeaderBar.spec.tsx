import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Button } from '@material-ui/core';
import HeaderBar, { Props } from '../../../app/components/headerBar/HeaderBar';
import GameManager from '../../../app/utils/GameManager';

Enzyme.configure({ adapter: new Adapter() });

const mockPush = spy();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockPush
	})
}));

const initialProps: Props = {
	gameManager: {} as GameManager
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<HeaderBar {...props} />);
	return {
		wrapper,
		button: wrapper.find(Button)
	};
};

describe('HeaderBar component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should have end game button', () => {
		const stopMock = spy();
		const props = { ...initialProps };
		props.gameManager = { stop: stopMock } as any;
		const { button } = setup(props);
		const mockPushCalls = mockPush.callCount;
		expect(button.first().text()).toMatch('End Game');
		button.first().simulate('click');
		expect(mockPush.calledWith('/')).toBeTruthy();
		expect(mockPush.callCount).toBe(mockPushCalls + 1);
		expect(stopMock.calledOnce).toBeTruthy();
	});

	it('should have miss button', () => {
		const missMock = spy();
		const props = { ...initialProps };
		props.gameManager = { miss: missMock } as any;
		const { button } = setup(props);
		expect(button.last().text()).toMatch('Miss');
		button.last().simulate('click');
		expect(missMock.calledOnce).toBeTruthy();
	});
});
