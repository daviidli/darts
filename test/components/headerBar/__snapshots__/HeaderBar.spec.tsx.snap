// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBar component should match snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="header"
  container={true}
  direction="row"
  justify="space-around"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={true}
  >
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
      variant="outlined"
    >
      End Game
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    item={true}
    justify="flex-end"
    xs={true}
  >
    <WithStyles(ForwardRef(Button))
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      Miss
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
